community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanUpdates.scala

```diff
@@ -31,40 +32,108 @@ case object PlanUpdates
+    case pattern: SetLabelPattern => context.logicalPlanProducer.planSetLabel(source, pattern)
     //SET n.prop = 42
     case pattern: SetNodePropertyPattern =>
-      context.logicalPlanProducer.planSetNodeProperty(inner, pattern)
+      context.logicalPlanProducer.planSetNodeProperty(source, pattern)
     //SET r.prop = 42
     case pattern: SetRelationshipPropertyPattern =>
-      context.logicalPlanProducer.planSetRelationshipProperty(inner, pattern)
+      context.logicalPlanProducer.planSetRelationshipProperty(source, pattern)
     //SET n.prop += {}
     case pattern: SetNodePropertiesFromMapPattern =>
-      context.logicalPlanProducer.planSetNodePropertiesFromMap(inner, pattern)
+      context.logicalPlanProducer.planSetNodePropertiesFromMap(source, pattern)
     //SET r.prop = 42
     case pattern: SetRelationshipPropertiesFromMapPattern =>
-      context.logicalPlanProducer.planSetRelationshipPropertiesFromMap(inner, pattern)
+      context.logicalPlanProducer.planSetRelationshipPropertiesFromMap(source, pattern)
     //REMOVE n:Foo:Bar
-    case pattern: RemoveLabelPattern => context.logicalPlanProducer.planRemoveLabel(inner, pattern)
+    case pattern: RemoveLabelPattern => context.logicalPlanProducer.planRemoveLabel(source, pattern)
     //DELETE a
     case p: DeleteExpression =>
       p.expression match {
         case Variable(n) if context.semanticTable.isNode(n) =>
-          context.logicalPlanProducer.planDeleteNode(inner, p)
+          context.logicalPlanProducer.planDeleteNode(source, p)
         case Variable(r) if context.semanticTable.isRelationship(r) =>
-          context.logicalPlanProducer.planDeleteRelationship(inner, p)
+          context.logicalPlanProducer.planDeleteRelationship(source, p)
         case PathExpression(e)  =>
-          context.logicalPlanProducer.planDeletePath(inner, p)
+          context.logicalPlanProducer.planDeletePath(source, p)
 
         case e => throw new CypherTypeException(s"Don't know how to delete a $e")
       }
   }
+
+  /*
+   * Merges either match or create. It is planned as following.
+   *
+   *                     |
+   *                anti-cond-apply
+   *                  /     \
+   *                 /    on-create
+   *                /        \
+   *               /   merge-create-part
+   *         cond-apply
+   *             /   \
+   *          apply  on-match
+   *          /   \
+   *         /  optional
+   *        /      \
+   *  (source)  merge-read-part
+   *
+   * Note also that merge uses a special leaf planner to enforce the correct behavior
+   * when having uniqueness constraints.
+   */
+  private def planMergeNode(query: PlannerQuery, source: LogicalPlan, merge: MergeNodePattern)(implicit context: LogicalPlanningContext) = {
+    //use a special unique-index leaf planner
+    val leafPlanners = PriorityLeafPlannerList(LeafPlannerList(mergeUniqueIndexSeekLeafPlanner),
```

Comment:
Would be nice to have a comment that explains the importance of forcing the use of unique indexes when they exist.